stackrox/stackrox
# ROX-10309: Add optional featureFlagDependency property to policy criteria descriptors

## Description

**Goal**: Minimize code changes to add and delete feature flags for policy criteria, therefore reduce merge conflicts between independent contributors.

### Changed files

Replace feature-specific with generic conditional logic.

1. Edit Containers/Policies/Wizard/Step3/BooleanPolicyLogicSection.tsx
    * Filter descriptors and replace array spread with ternary expression

2. Edit Containers/Policies/Wizard/Step3/PolicyCriteriaForm.tsx
    * Filter descriptors and replace array spread with ternary expression

3. Edit Containers/Policies/Wizard/Step3/policyCriteriaDescriptors.tsx
    * Add comment about `imageSigningCriteriaName` string and `signaturePolicyCriteria` type
    * Add `featureFlagDependency` property to `BaseDescriptor` type and 3 descriptors
    * Move 2 descriptors from `networkPolicyFieldDescriptors` to `policyConfigurationDescriptor`
    * Move `'Port Exposure'` descriptor to keep same order under **Networking** as with array spread
    * Move 1 descriptor from `imageSigningCriteriaDescriptor` to `policyConfigurationDescriptor`
    * Move 1 descriptor from `networkDetectionDescriptor` to `policyConfigurationDescriptor`

4. Edit types/featureFlag.ts
    * Add `'ROX_NETPOL_FIELDS'`

### Residue

1. Confirm correct naming convention for network policy criteria (for example, sentence case instead of title case).
2. Rename descriptor (singular) as descriptors (plural).
3. Encapsulate behavior of `signaturePolicyCriteria` type. Even better, as a more generic dynamic array type.

## Checklist

## Testing Performed

1. `yarn build` in ui
2. `yarn start` in ui
3. Visit /main/policies/?action=create
    * Select **Runtime** and **Audit logs** on step 2, and then see only **Kubernetes events** on step 3
    * Select **Build** on step 2, and then:
        See **Not verified by trusted image signers** under **Image registry** if `'ROX_VERIFY_IMAGE_SIGNATURE'` which is on by default
        ![ImageRegistry](https://user-images.githubusercontent.com/11862657/166237803-087ffbf4-2bdb-4643-83ab-cb7991040174.png)

        See **Ingress Network Policy** and **Egress Network Policy** under **Networking** if `'ROX_NETPOL_FIELDS'` which is off by default
        See **Unexpected network flow detected** under **Networking**
        ![Networking](https://user-images.githubusercontent.com/11862657/166237839-0636ca65-1ea4-4a77-8f09-6451ed1647da.png)
